### PR TITLE
lib: modem_slm: String null-terminator protection

### PR DESCRIPTION
**Pull request by junqingzou**
The strstr() function requires the source string to be null- terminated. It's risky to use to search in data buffer of UART reception as it could be partial of full AT response, meaning not null-terminated.

Solution is to search with length limit.

### CODE REVIEW

**Review by SeppoTakalo — CHANGES_REQUESTED** [review]

File: lib/modem_slm/modem_slm.c [hunk 1]
```diff
@@ -54,6 +54,34 @@ static const char at_usage_str[] = "Usage: slm <at_command>";
 /* global functions defined in different files */
 void slm_monitor_dispatch(const char *notif);
 
+/*
+ * __FBSDID("$FreeBSD: src/lib/libc/string/strnstr.c,v 1.5 2009/02/03 17:58:20 danger Exp $");
```
[the remark this answers is not in the document]
Comment (junqingzou) [reply lib/modem_slm/modem_slm.c:58]: How to?

**Comment by SeppoTakalo** [comment]
You only call this callback from one location
```c
	case UART_RX_RDY:
		LOG_DBG("RX_RDY %d", evt->data.rx.len);
		uart_rx_handler(&(evt->data.rx.buf[pos]), evt->data.rx.len);
```

Why don't you put the nul terminator here?

```c
	case UART_RX_RDY:
		LOG_DBG("RX_RDY %d", evt->data.rx.len);
                evt->data.rx.buf[evt->data.rx.offset + evt->data.rx.len] = 0;
		uart_rx_handler(&(evt->data.rx.buf[pos]), evt->data.rx.len);
```

Verify if that code works, before using it. I did not test it. Just a suggestion.

**Comment by junqingzou** [comment]
> You only call this callback from one location
> 
> ```c
> 	case UART_RX_RDY:
> 		LOG_DBG("RX_RDY %d", evt->data.rx.len);
> 		uart_rx_handler(&(evt->data.rx.buf[pos]), evt->data.rx.len);
> ```
> 
> Why don't you put the nul terminator here?
> 
> ```c
> 	case UART_RX_RDY:
> 		LOG_DBG("RX_RDY %d", evt->data.rx.len);
>                 evt->data.rx.buf[evt->data.rx.offset + evt->data.rx.len] = 0;
> 		uart_rx_handler(&(evt->data.rx.buf[pos]), evt->data.rx.len);
> ```
> 
> Verify if that code works, before using it. I did not test it. Just a suggestion.

No we cannot assume that there is always one extra byte in the buffer to add null-terminator.
For large response, on every CONFIG_MODEM_SLM_DMA_MAXLEN there will be one RX_RDY and the buffer is already full.

File: lib/modem_slm/modem_slm.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (junqingzou) [reply lib/modem_slm/modem_slm.c]: If there is extra room Zephyr Serial already set it to be 0x00.
See https://github.com/nrfconnect/sdk-zephyr/blob/main/drivers/serial/uart_nrfx_uarte.c#L1212

**Comment by SeppoTakalo** [comment]
> No we cannot assume that there is always one extra byte in the buffer to add null-terminator. For large response, on every CONFIG_MODEM_SLM_DMA_MAXLEN there will be one RX_RDY and the buffer is already full.

Then make a room:
```c
	case UART_RX_BUF_REQUEST:
		LOG_DBG("RX_BUF_REQ");
		pos = 0;
		err = uart_rx_buf_rsp(uart_dev, next_buf, sizeof(uart_rx_buf[0]) - 1);
		if (err) {
			LOG_WRN("UART RX buf rsp: %d", err);
		}
```

**Comment by junqingzou** [comment]
> > No we cannot assume that there is always one extra byte in the buffer to add null-terminator. For large response, on every CONFIG_MODEM_SLM_DMA_MAXLEN there will be one RX_RDY and the buffer is already full.
> 
> Then make a room:
> 
> ```c
> 	case UART_RX_BUF_REQUEST:
> 		LOG_DBG("RX_BUF_REQ");
> 		pos = 0;
> 		err = uart_rx_buf_rsp(uart_dev, next_buf, sizeof(uart_rx_buf[0]) - 1);
> 		if (err) {
> 			LOG_WRN("UART RX buf rsp: %d", err);
> 		}
> ```

This will mismatch DMA buffer size with SLM UART.
For example, SLM sends 256 bytes in one DMA and modem_slm will receive 255 + 1 in two DMA events.

**Comment by junqingzou** [comment]
Well, I'm not in favor of altering RX buffer so let me close this PR for a new one.
I'll forward the original bug report from customer as info, too.

**Comment by MarkusLassila** [comment]
> > You only call this callback from one location
> > ```c
> > 	case UART_RX_RDY:
> > 		LOG_DBG("RX_RDY %d", evt->data.rx.len);
> > 		uart_rx_handler(&(evt->data.rx.buf[pos]), evt->data.rx.len);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Why don't you put the nul terminator here?
> > ```c
> > 	case UART_RX_RDY:
> > 		LOG_DBG("RX_RDY %d", evt->data.rx.len);
> >                 evt->data.rx.buf[evt->data.rx.offset + evt->data.rx.len] = 0;
> > 		uart_rx_handler(&(evt->data.rx.buf[pos]), evt->data.rx.len);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Verify if that code works, before using it. I did not test it. Just a suggestion.
> 
> No we cannot assume that there is always one extra byte in the buffer to add null-terminator. For large response, on every CONFIG_MODEM_SLM_DMA_MAXLEN there will be one RX_RDY and the buffer is already full.

In principle, the response could contain multiple or partial responses or even something with 0 (like data in datamode), so I would not recommend adding 0 in it. If 0 was to be added, it should likely be added in uart_rx_handler to the ring buffer.

Generally, I don't think that this current implementation can handle the responses very well and there are situations where it will fail. If this needs to work in future as well, we should consider refactoring the UART implementation.

**Pull request closed without merging**